packages/core/src/extension/extension-messenger.ts

```diff
@@ -0,0 +1,108 @@
+import { IframeEventPoster } from '../common/iframe';
+import { appendWww, parseUrl } from '../common/url';
+import { bindDecorator as bind } from '../common/utility';
+
+import { ExtensionNotFoundError } from './errors';
+import { Extension } from './extension';
+import { ExtensionCommandHandlers } from './extension-command-handler';
+import { ExtensionPostEvent } from './extension-post-event';
+
+export class ExtensionMessenger {
+    _isListening = false;
+    _extensionCommandHandlers: ExtensionCommandHandlers | undefined;
+
+    constructor(
+        private _poster: IframeEventPoster<ExtensionPostEvent>,
+        private _extensions: Extension[] | undefined,
+    ) {}
+
+    listen(extensionCommandHandlers: ExtensionCommandHandlers): void {
+        this._extensionCommandHandlers = extensionCommandHandlers;
+
+        if (this._isListening) {
+            return;
+        }
+
+        this._isListening = true;
+
+        window.addEventListener('message', this._handleMessage);
+    }
+
+    post(event: ExtensionPostEvent, extensionId?: string): void {
+        if (extensionId && this._extensions) {
+            // Post to extension
+            const extension = this._extensions.find((e) => e.id === extensionId);
+
+            if (!extension) {
+                throw new ExtensionNotFoundError(
+                    `Unable to post due to no extension found for ID: ${extensionId}.`,
+                );
+            }
+
+            const iframe = document
+                .querySelector(`[data-extension-id="${extensionId}"]`)
+                ?.querySelector('iframe');
+
+            if (!iframe?.contentWindow) {
+                throw new ExtensionNotFoundError(
+                    `Unable to post due to no extension rendered for ID: ${extensionId}.`,
+                );
+            }
+
+            this._poster.setTarget(iframe.contentWindow);
+            this._poster.setTargetOrigin(extension.url);
+        } else {
```

Comment:
As I mentioned earlier, I think it's best to avoid changing the target. In this case, it's not very obvious that omitting the `extensionId` actually means posting to a different origin/target.

Reply:
Sure, now the post method is updated as below:
```
post(extensionId: string, event: ExtensionOriginEvent): void
```
`target` and `origin` are set at a poster's instantiation